index.d.ts

```diff
@@ -0,0 +1,33 @@
+import ESLint from 'eslint';
+
+declare const plugin: ESLint.ESLint.Plugin & {
+  meta: {
+    name: string;
+    version: string;
+  };
+  configs: {
+    'recommended': ESLint.Linter.LegacyConfig;
+    'errors': ESLint.Linter.LegacyConfig;
+    'warnings': ESLint.Linter.LegacyConfig;
+    'stage-0': ESLint.Linter.LegacyConfig;
+    'react': ESLint.Linter.LegacyConfig;
+    'react-native': ESLint.Linter.LegacyConfig;
+    'electron': ESLint.Linter.LegacyConfig;
+    'typescript': ESLint.Linter.LegacyConfig;
+  };
+  flatConfigs: {
+    'recommended': ESLint.Linter.LegacyConfig;
+    'errors': ESLint.Linter.LegacyConfig;
+    'warnings': ESLint.Linter.LegacyConfig;
+    'stage-0': ESLint.Linter.LegacyConfig;
+    'react': ESLint.Linter.LegacyConfig;
+    'react-native': ESLint.Linter.LegacyConfig;
+    'electron': ESLint.Linter.LegacyConfig;
+    'typescript': ESLint.Linter.LegacyConfig;
+  };
+  rules: {
+    [key: string]: ESLint.Rule.RuleModule;
+  };
```

Comment:
We could also add types for `settings`, something like:

```js
settings: Record<string, unknown> & {
  // we can add there any known property
  'import/resolver'?: Record<string, unknown>
  typescript?: { alwaysTryTypes?: boolean }
  'extensions'?: ...
}
```

Reply:
Yup in the long-run that'd be ideally and I'm happy to help work on that but I'd prefer to not block this on that since I'm not super familiar with all the settings and really right now their types won't get used out of the box since the new config system doesn't result in the types getting automatically applied.

i.e. they'd be useful for letting you manually type an object via `pluginImport['settings']` but that's a more complex setup

Reply:
(I actually only happened to include `rules` at all as I had that type on hand from what we've got in `eslint-plugin-jest` - you could make the same point that ideally the rules could be typed with all their settings, but again that'd be a lot of work for something that right now wouldn't be that usable)